core/src/main/java/net/opentsdb/query/execution/serdes/JsonV2ExpQuerySerdes.java

```diff
@@ -389,6 +349,147 @@ public Object call(final Exception ex) throws Exception {
+          }
+        } else if (value.type() == NumericSummaryType.TYPE) {
+          final Collection<Integer> summaries = 
+              ((TimeSeriesValue<NumericSummaryType>) value).value()
+              .summariesAvailable();
+          if (!summaries.isEmpty()) {
+            // JUST grab the first
+            int summary = summaries.iterator().next();
+            final NumericType v = ((TimeSeriesValue<NumericSummaryType>) value)
+                .value().value(summary);
+            if (v == null) {
+              // TODO - actual null or keep it NaN?
+              json.writeNumber(Double.NaN);
+            } else if (v.isInteger()) {
+              json.writeNumber(v.longValue());
+            } else {
+              json.writeNumber(v.doubleValue());
+            }
+          } else {
+            json.writeNumber(Double.NaN);
+          }
+        }
+        
+        if (iterators.get(i).hasNext()) {
+          has_next = true;
+          next_next_ts.update(interpolator.nextReal());
+        }
+      }
+      
+      json.writeEndArray();
+      if (has_next) {
+        if (next_ts.compare(Op.EQ, next_next_ts)) {
+          LOG.error("Unexpected exception: Same timestamp as before: " + next_ts);
+          return set_count;
+        }
+        next_ts.update(next_next_ts);
+      } else {
+        return set_count;
+      }
+
+      if (next_ts.compare(Op.GTE, context.query().endTime())) {
+        return set_count;
+      }
+    }
+    
+    return set_count;
+  }
+
+  int serializeArray(final QueryResult result, 
+                     final List<TypedTimeSeriesIterator<?>> iterators) 
+                         throws IOException {
+    int index = 0;
+    final TimeStamp next_ts = result.timeSpecification().start().getCopy();
+    
+    while (next_ts.compare(Op.LT, context.query().startTime())) {
+      index++;
+      next_ts.add(result.timeSpecification().interval());
+    }
+    
+    final TimeSeriesValue[] values = new TimeSeriesValue[iterators.size()];
+    for (int i = 0; i < iterators.size(); i++) {
+      values[i] = iterators.get(i).next();
+    }
+    
+    while (true) {
+      json.writeStartArray();
+      json.writeNumber(next_ts.msEpoch());
+      
+      for (int i = 0; i < values.length; i++) {
+        final NumericArrayType value = 
+            ((TimeSeriesValue<NumericArrayType>) values[i]).value();
+        
+        if (value.offset() + index >= value.end()) {
+          // done!
+          return iterators.size();
+        }
+        
+        if (value.isInteger()) {
+          json.writeNumber(value.longArray()[value.offset() + index]);
+        } else {
+          json.writeNumber(value.doubleArray()[value.offset() + index]);
+        }
+      }
+    
+      index++;
+      next_ts.add(result.timeSpecification().interval());
+      json.writeEndArray();
+      if (next_ts.compare(Op.GTE, context.query().endTime())) {
```

Comment:
If the query endtime is in the future, the next_ts will never be greater than query endtime and will it cause an infi loop?

Reply:
Ah good catch, I missed the has_next call in here.

Reply:
Actually the other catch is on L473 where it will bail when we reach the end of the array.